### DISCUSSION TITLE
Codebase health

### DISCUSSION THREAD

**Discussion by enrichman** [epinio/epinio#1249] [General]
This is a broad discussion about how we could improve the health of the Epinio codebase. 🧑‍⚕️

There are different topics that came into my mind:
- tests
- linters
- project structure
- interfaces

I think we can brainstorm a bit on the various topics and maybe open a dedicated discussion for each one if we need to.

## Tests
We have a lot of acceptance/end-to-end tests, and this is good, since we have lot of moving pieces, cloud providers, and so on. But we have very few unit tests. While I agree that code coverage is not a useful metrics, I have to say that writing tests is a good way on writing good code.

About this I would propose to add a coverage badge and a goal that we should aim to. At the moment we have a test coverage of less than **20%**. The final goal would be at **80%**, but even if we could reach the 50% I think we could be happy about it!

We could embed some external tools like [Codecov](https://about.codecov.io/) or [Coveralls](https://coveralls.io/) to check this.

## Linters
We have added the new `funlen` linter to check for too lengthy function. We should lower the threshold and arrive (hopefully) to the default value.

## Project structure
This topic is going to be _highly_ opinionated, but I think we can do better on structuring the project.
There is no golden rule on how to structure a Go project, but these are my 2¢, while trying to find [resources](https://github.com/golang-standards/project-layout) and guidelines.

I have the feeling that having a no clear guideline on where to put things had brought us to an unclear situation. One "code smell" that I feel is the `helpers` package that looks too big to me, maybe there are things that could be moved in a more appropriate position. For example the `kubernetes` package looks pretty big, with lot of things. Would it make sense to move it inside the `internal` package?

## Interfaces
We have very few interfaces, and we usually pass structs to the functions. This is going to couple the code a lot, and make it really hard to test. We should avoid this.

These are just some small things that came up into my mind, feel free to add more, and let me know. I could missing some context!

**Comment by andreas-kupries** [comment]
> the helpers 

directory is intended to contain the packages which can, conceivably, used outside of the epinio project itself.

OTOH `internal` is for the packages which have no use outside of epinio itself.
This has been weakened somewhat with the `pkg/` hierarchy containing mostly cross-package types (models, etc.).
The are `internal` as well.

**Reply by enrichman** [reply]
Well, as you said that should be the [main purpose](https://github.com/golang-standards/project-layout#pkg) of the `/pkg` folder. We should probably move some of these things there.

We have [some](https://github.com/epinio/epinio/blob/main/helpers/authtoken/token.go#L50-L56) Epinio related stuff in the `helpers` package, so we should abstract a bit their use, or move them. Anyway this is the most difficult topic. :smile:
